lib/presentation/manager/upload_and_share_file/upload_and_share_file_manager.dart

```diff
@@ -0,0 +1,153 @@
+import 'package:redux/redux.dart';
+
+class UploadShareFileManager {
+  UploadShareFileManager(
+    this._store,
+    this._uploadFileInteractor,
+    this._shareDocumentInteractor,
+  );
+
+  final Store<AppState> _store;
+
+  final UploadFileInteractor _uploadFileInteractor;
+  final ShareDocumentInteractor _shareDocumentInteractor;
+
+  void justUpload(FileInfo uploadFile) async {
+    await _upload(uploadFile).forEach((uploadState) {
+      _store.dispatch(UploadFileAction(uploadState));
+    });
+  }
+
+  void uploadThenShare(
+    FileInfo uploadFile,
+    List<AutoCompleteResult> recipients,
+  ) async {
+    await _upload(uploadFile).forEach((uploadState) {
+      uploadState.fold(
+        (failure) => _store.dispatch(UploadFileAction(uploadState)),
+        (success) async {
+          if (success is FileUploadSuccess) {
+            await _shareAfterUploaded(success.uploadedDocument, recipients);
+          } else {
+            _store.dispatch(UploadFileAction(uploadState));
+          }
+        },
+      );
+    });
+  }
+
+  void _shareAfterUploaded(
+    Document uploadedDocument,
+    List<AutoCompleteResult> recipients,
+  ) async {
+    _store.dispatch(UploadFileAction(Right(SharingAfterUploadState(recipients, uploadedDocument))));
+
+    final shareResult = await _share(recipients, [uploadedDocument.documentId]);
+
+    shareResult.fold(
+      (failure) => _store.dispatch(
+        ShareAction(Left(ShareAfterUploadFailure(Exception(), recipients, uploadedDocument))),
+      ),
+      (success) => _store.dispatch(
+        ShareAction(Right(ShareAfterUploadSuccess(recipients, uploadedDocument))),
+      ),
+    );
+  }
+
+  void justShare(
+    List<AutoCompleteResult> recipients,
+    List<DocumentId> shareDocuments,
+  ) async {
+    await _share(recipients, shareDocuments)
+        .then((shareResult) => _store.dispatch(ShareAction(shareResult)));
+  }
+
+  Stream<Either<Failure, Success>> _upload(FileInfo uploadFile) {
+    return _uploadFileInteractor.execute(uploadFile);
+  }
+
+  Future<Either<Failure, Success>> _share(
+    List<AutoCompleteResult> recipients,
+    List<DocumentId> shareDocuments,
+  ) async {
+    final genericUsers = recipients
+        .where(
+          (element) =>
+              element is UserAutoCompleteResult ||
+              element is SimpleAutoCompleteResult,
+        )
+        .map(
+          (data) => GenericUser(
+            _getSuggestionMail(data),
+            firstName: none(),
+            lastName: none(),
+          ),
+        )
+        .toList();
+    final mailingListIds = recipients
+        .whereType<MailingListAutoCompleteResult>()
+        .map((data) => MailingListId(data.identifier))
+        .toList();
+
+    return await _shareDocumentInteractor.execute(
+      shareDocuments,
+      mailingListIds,
+      genericUsers,
+    );
+  }
+
+  String _getSuggestionMail(AutoCompleteResult autoCompleteResult) {
```

Comment:
Let me define a common method for this in my PR, you can rebase to use it later 👍